packages/financial-templates-lib/src/clients/InsuredBridgeL1Client.ts

```diff
@@ -299,6 +369,45 @@ export class InsuredBridgeL1Client {
       this.instantRelays[l1Token] = {};
     }
 
+    // Fetch and store all rate model updated events, which the user of this client can use to fetch a rate model for a
+    // specific deposit quote timestamp.
+    const updatedRateModelEvents = await this.rateModelStore.getPastEvents("UpdatedRateModel", blockSearchConfig);
+    for (const updatedRateModelEvent of updatedRateModelEvents) {
+      // The contract enforces that all rate models are mapped to addresses, therefore we do not need to check that
+      // `l1Token` is a valid address.
+      const l1TokenNormalized = toChecksumAddress(updatedRateModelEvent.returnValues.l1Token);
+      if (!this.updatedRateModelEventsForToken[l1TokenNormalized])
+        this.updatedRateModelEventsForToken[l1TokenNormalized] = [];
+
+      // Before adding the new rate model to the array, first validate its format:
+      let rateModelFromEvent: any;
+      try {
+        rateModelFromEvent = JSON.parse(updatedRateModelEvent.returnValues.rateModel);
+      } catch (err) {
+        // Found an UpdatedRateModel event where the rate model string cannot be parsed into JSON, ignore it.
+        continue;
+      }
+
+      if (!expectedRateModelKeys.every((item) => Object.prototype.hasOwnProperty.call(rateModelFromEvent, item))) {
```

Comment:
A few questions:
1. Should an improperly formatted rate model throw an error? What happens if there's a rate model that matches our expectations at block 1, it's updated to one that has "incorrect" formatting at block 2, and then we're trying to find a rate model at block 3? Do we just drop the update and return the stale one? That seems undesirable.
2. nit: this code seems a bit convoluted. Could we just use `in`? ex: `item => item in rateModelFromEvent`?
3. Do we care if there are extra properties in the rateModelFromEvent? i.e. should we check the reverse, that each property in rateModelFromEvent is in the `expectedRateModelKeys` array?

Reply:
1. Hm this is a great point, and one that can be resolved by lazily parsing stringified rate models. I've changed this to store the stringified rate model events in `update`, and then only attempt to lazily parse it in `getRateModelForBlockNumber` when it needs to be used. At that point, a failure to parse will throw an error. This way if an incorrect rate model is inserted in block 2, the rate model at block 3 will be this incorrect rate model and parsing it will throw an error.
2. Good point done.
3. I believe this check already allows for extra keys. It takes every key in `expectedRateModelKeys` and searches for it in `rateModelFromEvent`. Do you understand otherwise?